rule/unsecure_url_scheme.go

```diff
@@ -0,0 +1,71 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// UnsecureURLSchemeRule checks if a file contains string literals with unsecure URL schemes (for example: http://... in place of https://...).
+type UnsecureURLSchemeRule struct{}
+
+// Apply applied the rule to the given file.
+func (r *UnsecureURLSchemeRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	if file.IsTest() {
+		return nil // skip test files
+	}
+
+	var failures []lint.Failure
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	w := lintUnsecureURLSchemeRule{
+		onFailure: onFailure,
+	}
+
+	ast.Walk(w, file.AST)
+	return failures
+}
+
+// Name returns the rule name.
+func (*UnsecureURLSchemeRule) Name() string {
+	return "unsecure-url-scheme"
+}
+
+type lintUnsecureURLSchemeRule struct {
+	onFailure func(lint.Failure)
+}
+
+func (w lintUnsecureURLSchemeRule) Visit(node ast.Node) ast.Visitor {
+	n, ok := node.(*ast.BasicLit)
+	if !ok || n.Kind != token.STRING {
+		return w // not a string litereal
+	}
+
+	value := n.Value
+	var scheme string
+	switch {
+	case strings.HasPrefix(value, `"http://`):
+		scheme = "http"
+	case strings.HasPrefix(value, `"ws://`):
+		scheme = "ws"
+	default:
+		return nil // not an URL or not an unsecure one
+	}
+
+	if strings.Contains(value, "localhost") || strings.Contains(value, "127.0.0.1") {
```

Comment:
Do we have example of `http://::1` or `http://::/`

Here I feel like we should detect the presence of http:// and ws:// and then parse URL, but this might not work for string concatenation

Reply:
In the first draft of the rule I've used a regexp to deconstruct the URL but I felt the approach overkilling. By "simply" using HasPrefix and Contains the rule copes with the generality of cases and keeps its performance acceptable.